include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h

```diff
@@ -63,6 +76,40 @@ class AnnotationSet {
   DictionaryAttr
   getArgumentAttrDict(ArrayRef<NamedAttribute> otherPortAttrs = {}) const;
 
+  /// Store this annotation set on an operation. If `removeOnEmpty` is set, the
+  /// annotation attribute is removed from the operation altogether. Returns
+  /// true if the operation was modified, false otherwise.
+  bool updateOperation(Operation *op, bool removeOnEmpty = true) const;
```

Comment:
The `removeOnEmpty` parameter seems like a bit of a weird option. It seems like there should be a policy of "everything has an `"annotations"` attribute" or "empty annotation attributes aren't stored". The `AnnotationSet` constructor will already do the right thing when `"annotations"` doesn't exist, so maybe it's fine to remove `removeOnEmpty` and make this the policy for how this is handled?

Reply:
How about `annotSet.applyToOperation(op);`?

Out of curiosity, why do we need the `removeOnEmpty` parameter?  Isn't this a function of the operation, so this could just do the right thing depending on what the operation is?

Reply:
Nice, I also prefer to get rid of the `removeOnEmpty` parameter and remove empty `annotations = {}` attributes. Otherwise we'll end up with a lot of these dead entries around. Just left the parameter there to leave this to the user, but actually enforcing this behaviour as part of the `AnnotationSet` API is even better!